File: VBFNLO_eemm_Raw.py
#!/usr/bin/python

##############
## Make Plots
##############

########################
# Libraries and Includes
########################

# Python System Modules
import os,sys,glob
import logging

import array
# Python Paths
ROOTLIB, PATH_OnePlot_MODULE1, PATH_OnePlot_MODULE2 = "", "", ""

if os.getenv("ROOTLIB"): ROOTLIB=os.getenv("ROOTLIB")
if os.getenv("PATH_OnePlot_MODULE1"): PATH_OnePlot_MODULE1=os.getenv("PATH_OnePlot_MODULE1")
if os.getenv("PATH_OnePlot_MODULE2"): PATH_OnePlot_MODULE2=os.getenv("PATH_OnePlot_MODULE2")

sys.path.append(ROOTLIB)
sys.path.append(PATH_OnePlot_MODULE1)
sys.path.append(PATH_OnePlot_MODULE2)

# Python Math Modules
import array
from math import sqrt,fabs,sin,pow

# ROOT Modules
from ROOT import TFile,TTree,TChain,TBranch,TH1F,TH2F
from ROOT import TCanvas,TPad,TLine
from ROOT import TLorentzVector
from ROOT import gROOT, gDirectory,gStyle
from ROOT import TPaveText, TLegend 
import ROOT

# user modules
from oneplot import oneplot
from module_syst import systPDFCT10, systEnvelope, syst1v1, systCombine
from module_style import atlas_style
    
###############################
# User defined drawing function
###############################

if __name__ == "__main__":

    ############################
    # Plot for Unfolding Results
    ############################
    
    files="eemm_185019_fid_hist.root"
    
    # open file
    fin = TFile(files)

    VAR_4 = [ [ ["l"], ["e", "pt", "eta", "phi"] ] ]
    VAR_2 = [ [ ["z"], ["pt", "eta", "phi", "m"] ] ] 
    VAR_1 = [ [ ["zz", "jj"], [ "m", "pt", "eta", "phi"] ] ]
    VAR_0 = ["delta_R_zz", "delta_R_jj", "delta_R_zz_jj", "mcweight", "delta_eta_jj", "centrality"]
    VAR_B = [("delta_R_ll_1", "delta_R_ll_2")] 
    VAR_J = ["pt", "eta", "phi"]

    TITLES = {"l_e":["E [GeV]","events / 10GeV "],
              "l_pt":["Pt [Gev]","events / 10GeV "],
              "l_eta":["#eta","events / 0.35 "],
              "l_phi":["#phi","events / 0.35 "],
              "z_pt":["Pt [Gev]","events / 10GeV "], 
              "z_eta":["#eta","events / 0.5 "],      
              "z_phi":["#phi","events / 0.35 "],      
              "z_m":["M [GeV]", "events / 6GeV"],
              "z_e":["E [GeV]","events / 20GeV "],   
              "leadingj_e-subleadingj_e":["E [GeV]","events / 20GeV "],   
              "leadingj_pt-subleadingj_pt":["Pt [Gev]","events / 10GeV "], 
              "leadingj_eta-subleadingj_eta":["#eta","events / 0.5 "],      
              "leadingj_phi-subleadingj_phi":["#phi","events / 0.35 "],      
              "zz_pt":["Pt [Gev]","events / 10GeV "], 
              "zz_eta":["#eta","events / 0.2 "],      
              "zz_phi":["#phi","events / 0.2 "],      
              "zz_m":["M [GeV]", "events / GeV"],
              "zz_e":["E [GeV]","events / 20GeV "],   
              "jj_pt":["Pt [Gev]","events / 10GeV "], 
              "jj_eta":["#eta","events / 0.2 "],      
              "jj_phi":["#phi","events / 0.2 "],      
              "jj_m":["M [GeV]", "events / 60GeV"],
              "jj_e":["E [GeV]","events / 20GeV "],   
              "delta_R_zz":["#{Delta}R_zz", "events/ 0.5"],
              "delta_R_jj":["#{Delta}R_jj", "events/ 0.5"],
              "delta_R_zz_jj":["#{Delta}R_zzjj", "events/ 0.5"],
              "delta_R_ll_1-delta_R_ll_2":["#{Delta}R_ll", "events/ 0.5"],
              "mcweight":["weight", "events/ 0.5"],
              "delta_eta_jj":["#Delta#eta_{jj}","events / 0.1"],
              "centrality":["centrality","events / 0.1 "]
              }
    LEGENDS = { 
                "deltaEta_jj":"#Delta#eta",
                "centrality":"centrality"
              }

    for entry in VAR_4:
        for par in entry[0]:
            for var in entry[1]:

                # load histograms
                histos = []
                names = []
                var_name = par + "_" + var
                for i in range(1,5):
                    hist_name = par + str(i) + "_" + var
                    names.append(hist_name)
                    histo_tmp = fin.Get(hist_name)
                    histos.append(histo_tmp)

                # Draw histograms
                xtitle = TITLES[var_name][0]
                ytitle = TITLES[var_name][1]
                legends = ["e1", "e2", "m1", "m2"]
                options = ["HIST", "HIST", "HIST", "HIST"]            
                opt_legends = ["L", "L", "L", "L"]                    
                marker_types = [26, 26, 32, 32]                       
                marker_sizes = [1, 1, 1, 1]                           
                line_sizes = [2, 2, 2, 2]                             
                marker_colors=line_colors=fill_colors=[4, 6, 2, 12] 
                fill_types = [0, 0, 0, 0]                             
                figname = var_name
        
                # call one plot
                theone = oneplot()
                theone._figtype=["png"]
                theone.initialize(list_histo=histos,names=names,legends=legends, opt_legends=opt_legends,
                                xtitle=xtitle,ytitle=ytitle,
                                figname=figname, 
                                options=options, marker_types=marker_types, marker_colors=marker_colors, marker_sizes=marker_sizes,
                                line_colors=line_colors, line_sizes=line_sizes,
                                fill_colors=fill_colors, fill_types=fill_types)
                theone.plot1DHistogram()
                theone.finish()

    for entry in VAR_2:
        for par in entry[0]:
            for var in entry[1]:

                # load histograms
                histos = []
                names = []
                var_name = par + "_" + var
                for i in range(1,3):
                    hist_name = par + str(i) + "_" + var
                    names.append(hist_name)
                    histo_tmp = fin.Get(hist_name)
                    histos.append(histo_tmp)

                # Draw histograms
                xtitle = TITLES[var_name][0]
                ytitle = TITLES[var_name][1]
                legends = ["z_ee", "z_mm"]
                options = ["HIST", "HIST"]            
                opt_legends = ["L", "L"]                    
                marker_types = [26, 32]                       
                marker_sizes = [1, 1]                           
                line_sizes = [2, 2]                             
                marker_colors=line_colors=fill_colors=[4, 6] 
                fill_types = [0, 0]                             
                figname = var_name
        
                # call one plot
                theone = oneplot()
                theone._figtype=["png"]
                theone.initialize(list_histo=histos,names=names,legends=legends, opt_legends=opt_legends,
                                xtitle=xtitle,ytitle=ytitle,
                                figname=figname, 
                                options=options, marker_types=marker_types, marker_colors=marker_colors, marker_sizes=marker_sizes,
                                line_colors=line_colors, line_sizes=line_sizes,
                                fill_colors=fill_colors, fill_types=fill_types)
                theone.plot1DHistogram()
                theone.finish()


    for entry in VAR_1:
        for par in entry[0]:
            for var in entry[1]:
                histo_name = par + "_" + var
                names = [histo_name]
                histos = []

                # load histograms
                histo_tmp = fin.Get(histo_name)
                histos.append(histo_tmp)

                # Draw histograms
                xtitle = TITLES[histo_name][0]
                ytitle = TITLES[histo_name][1]
                legends = [histo_name]
                options = ["HIST"]
                opt_legends = ["L"]
                marker_types = [26]
                marker_sizes = [1]
                line_sizes = [2]
                marker_colors=line_colors=fill_colors=[4]
                fill_types = [0]
                figname=histo_name
        
                # call one plot
                theone = oneplot()
                theone._figtype = ["png"]
                theone.initialize(list_histo=histos,names=names,legends=legends, opt_legends=opt_legends,
                                xtitle=xtitle,ytitle=ytitle,
                                figname=figname, 
                                options=options, marker_types=marker_types, marker_colors=marker_colors, marker_sizes=marker_sizes,
                                line_colors=line_colors, line_sizes=line_sizes,
                                fill_colors=fill_colors, fill_types=fill_types)
                newline = TLine(0.2,0.2,0.4,0.4)
                newline.Draw()
                theone.plot1DHistogram()
                theone.finish()

    for var in VAR_0:
        histo_name = var
        names = [var]
        histos = []

        # load histograms
        histo_tmp = fin.Get(histo_name)
        histos.append(histo_tmp)

        # Draw histograms
        xtitle = TITLES[histo_name][0]
        ytitle = TITLES[histo_name][1]
        legends = [var]
        options = ["HIST"]
        opt_legends = ["L"]
        marker_types = [26]
        marker_sizes = [1]
        line_sizes = [2]
        marker_colors=line_colors=fill_colors=[4]
        fill_types = [0]
        figname=histo_name

        # call one plot
        theone = oneplot()
        theone._figtype = ["png"]
        theone.initialize(list_histo=histos,names=names,legends=legends, opt_legends=opt_legends,
                        xtitle=xtitle,ytitle=ytitle,
                        figname=figname, 
                        options=options, marker_types=marker_types, marker_colors=marker_colors, marker_sizes=marker_sizes,
                        line_colors=line_colors, line_sizes=line_sizes,
                        fill_colors=fill_colors, fill_types=fill_types)
        newline = TLine(0.2,0.2,0.4,0.4)
        newline.Draw()
        theone.plot1DHistogram()
        theone.finish()

    for var in VAR_B:
        var1, var2 = var
        names = [var1, var2]
        histos = []

        histo_tmp = fin.Get(var1)
        histos.append(histo_tmp)
        histo_tmp = fin.Get(var2)
        histos.append(histo_tmp)

        xtitle = TITLES[var1+"-"+var2][0]
        ytitle = TITLES[var1+"-"+var2][1]
        var1.replace("1", "ee")
        var2.replace("2", "mm")
        legends = [var1, var2]
        options = ["HIST", "HIST"]            
        opt_legends = ["L", "L"]                    
        marker_types = [26, 32]                       
        marker_sizes = [1, 1]                           
        line_sizes = [2, 2]                             
        marker_colors=line_colors=fill_colors=[4, 6] 
        fill_types = [0, 0]                             
        figname = var1 + var2

        # call one plot
        theone = oneplot()
        theone._figtype=["png"]
        theone.initialize(list_histo=histos,names=names,legends=legends, opt_legends=opt_legends,
                        xtitle=xtitle,ytitle=ytitle,
                        figname=figname, 
                        options=options, marker_types=marker_types, marker_colors=marker_colors, marker_sizes=marker_sizes,
                        line_colors=line_colors, line_sizes=line_sizes,
                        fill_colors=fill_colors, fill_types=fill_types)
        theone.plot1DHistogram()
        theone.finish()
    
    for var in VAR_J:
        var1 = "leadingj_" + var
        var2 = "subleadingj_" + var
        names = [var1, var2]
        histos = []

        histo_tmp = fin.Get(var1)
        histos.append(histo_tmp)
        histo_tmp = fin.Get(var2)
        histos.append(histo_tmp)

        xtitle = TITLES[var1+"-"+var2][0]
        ytitle = TITLES[var1+"-"+var2][1]
        legends = ["j1", "j2"]
        options = ["HIST", "HIST"]            
        opt_legends = ["L", "L"]                    
        marker_types = [26, 32]                       
        marker_sizes = [1, 1]                           
        line_sizes = [2, 2]                             
        marker_colors=line_colors=fill_colors=[4, 6] 
        fill_types = [0, 0]                             
        figname = "jet_" + var

        # call one plot
        theone = oneplot()
        theone._figtype=["png"]
        theone.initialize(list_histo=histos,names=names,legends=legends, opt_legends=opt_legends,
                        xtitle=xtitle,ytitle=ytitle,
                        figname=figname, 
                        options=options, marker_types=marker_types, marker_colors=marker_colors, marker_sizes=marker_sizes,
                        line_colors=line_colors, line_sizes=line_sizes,
                        fill_colors=fill_colors, fill_types=fill_types)
        theone.plot1DHistogram()
        theone.finish()
    # plot test 1 (plot all truth distribution together)
    #names=["nom", "PSUp", "PSDown", "ScaleUp", "ScaleDown", "PDFUp", "PDFDown", "EWKUp", "EWKDown"]
    #legends=["nom", "PSUp", "PSDown", "ScaleUp", "ScaleDown", "PDFUp", "PDFDown", "EWKUp", "EWKDown"]
    #histos=[histo_mc, histo_mc_psup, histo_mc_psdn, histo_mc_scaleup, histo_mc_scaledn, histo_mc_pdfup, histo_mc_pdfdn, histo_mc_ewkup, histo_mc_ewkdn]
    #figname="compare_mctruth_{0}".format(keyname)
    #options=["HIST", "LPE", "LPE", "LPE", "LPE", "LPE", "LPE", "LPE", "LPE"]
    #marker_types=[1, 1, 1, 1, 1, 1, 1, 1, 1]
    #line_colors=[1, 2, 2, 3, 3, 4, 4, 5, 5]
    #line_sizes=[2, 2, 2, 2, 2, 2, 2, 2, 2]
    #theone = oneplot()
    #theone.initialize(list_histo=histos,names=names,legends=legends,
    #                ratio=0.2,xtitle="{0} [GeV]".format(xname),ytitle="#Delta#sigma / Bin [fb]",
    #                figname=figname, ratiotitle="Ratio",
    #                equalbin=True, options=options, line_colors=line_colors, line_sizes=line_sizes)
    #theone.plot1DHistogram()
    #theone.finish()
    
    fin.Close()
    
    # save the files into histograms
    #fout = TFile(figname+".root", "recreate")
    #histo_unfold_final.Write()
    #histo_mc_final.Write()



    #equalbin=True, fout.Close()

File: VBFNLO_eemm_Raw.cxx
#include <iostream>
#include <string>
#include <time.h>
#include "TLorentzVector.h"
#include "TFile.h"
#include "TTree.h"
using namespace std;

// dealing with VBFNLO eemm channel
// variables: l1,l2 --> e1,e2 (e1.pt > e2.pt)
//	      l3,l4 --> m1,m2 (m1.pt > m2.pt)
//	      z1,z2 --> z_ee, z_mm
//	      delta_R_ll_1,delta_R_ll_2 --> delta_R_ll_ee,delta_R_ll_mm

static double Z_MASS = 91.1876; // unit in GeV 
time_t start, finish;

void raw(string input_file, string output_file) 
{
    time(&start);

    TFile * fin = new TFile(input_file.c_str(), "read");
    if( ! fin ) 
    {
	cout << "Error, file " << input_file << "doesn't exit" << endl;
	return ;
    }

    TTree * tin = (TTree *) fin->Get("vbfnlo");
    if ( ! tin ) 
    {
	cout << "tin vbfnlo doesn't exist in file " << input_file << endl;
	return ;
    }

    TFile * fout = new TFile(output_file.c_str(), "recreate");
    if(! fout)
    {
	cout << "Error! unable to create file: " << output_file << endl;
	return;
    }

    TTree * tout = new TTree("vbfnlo", "vbfnlo");
    double px[10];
    double py[10];
    double pz[10];
    double E[10];
    double M[10];
    int pdgID[10];
    int status[10];
    int parent1_index[10];
    int parent2_index[10];
    double weight;
    tin->SetBranchAddress("mc_px", px);
    tin->SetBranchAddress("mc_py", py);
    tin->SetBranchAddress("mc_pz", pz);
    tin->SetBranchAddress("mc_E", E);
    tin->SetBranchAddress("mc_m", M);
    tin->SetBranchAddress("mc_pdgID", pdgID);
    tin->SetBranchAddress("mc_status", status);
    tin->SetBranchAddress("mc_parent1_index", parent1_index);
    tin->SetBranchAddress("mc_parent2_index", parent2_index);
    tin->SetBranchAddress("mc_evtweight", &weight);

    float l1_pt, l1_eta, l1_phi, l1_e;
    float l2_pt, l2_eta, l2_phi, l2_e;
    float l3_pt, l3_eta, l3_phi, l3_e;
    float l4_pt, l4_eta, l4_phi, l4_e;
    float leadingj_pt, leadingj_eta, leadingj_phi, leadingj_e;
    float subleadingj_pt, subleadingj_eta, subleadingj_phi, subleadingj_e;
    float z1_pt, z1_eta, z1_phi, z1_e, z1_m;
    float z2_pt, z2_eta, z2_phi, z2_e, z2_m;
    float zz_pt, zz_eta, zz_phi, zz_e, zz_m;
    float jj_pt, jj_eta, jj_phi, jj_e, jj_m;
    float delta_eta_jj, centrality;
    float delta_R_ll_1, delta_R_ll_2, delta_R_zz, delta_R_jj, delta_R_zz_jj;
    tout->Branch("l1_pt",  &l1_pt);
    tout->Branch("l1_eta", &l1_eta);
    tout->Branch("l1_phi", &l1_phi);
    tout->Branch("l1_e",   &l1_e);
    tout->Branch("l2_pt",  &l2_pt);
    tout->Branch("l2_eta", &l2_eta);
    tout->Branch("l2_phi", &l2_phi);
    tout->Branch("l2_e",   &l2_e);
    tout->Branch("l3_pt",  &l3_pt);
    tout->Branch("l3_eta", &l3_eta);
    tout->Branch("l3_phi", &l3_phi);
    tout->Branch("l3_e",   &l3_e);
    tout->Branch("l4_pt",  &l4_pt);
    tout->Branch("l4_eta", &l4_eta);
    tout->Branch("l4_phi", &l4_phi);
    tout->Branch("l4_e",   &l4_e);
    tout->Branch("leadingj_pt",  &leadingj_pt);
    tout->Branch("leadingj_eta", &leadingj_eta);
    tout->Branch("leadingj_phi", &leadingj_phi);
    tout->Branch("leadingj_e",   &leadingj_e);
    tout->Branch("subleadingj_pt",  &subleadingj_pt);
    tout->Branch("subleadingj_eta", &subleadingj_eta);
    tout->Branch("subleadingj_phi", &subleadingj_phi);
    tout->Branch("subleadingj_e",   &subleadingj_e);
    tout->Branch("z1_pt",  &z1_pt);
    tout->Branch("z1_eta", &z1_eta);
    tout->Branch("z1_phi", &z1_phi);
    tout->Branch("z1_e",   &z1_e);
    tout->Branch("z1_m",   &z1_m);
    tout->Branch("z2_pt",  &z2_pt);
    tout->Branch("z2_eta", &z2_eta);
    tout->Branch("z2_phi", &z2_phi);
    tout->Branch("z2_e",   &z2_e);
    tout->Branch("z2_m",   &z2_m);
    tout->Branch("zz_pt",  &zz_pt);
    tout->Branch("zz_eta", &zz_eta);
    tout->Branch("zz_phi", &zz_phi);
    tout->Branch("zz_e",   &zz_e);
    tout->Branch("zz_m",   &zz_m);
    tout->Branch("jj_pt",  &jj_pt);
    tout->Branch("jj_eta", &jj_eta);
    tout->Branch("jj_phi", &jj_phi);
    tout->Branch("jj_e",   &jj_e);
    tout->Branch("jj_m",   &jj_m);
    tout->Branch("delta_R_ll_1",   &delta_R_ll_1);
    tout->Branch("delta_R_ll_2",   &delta_R_ll_2);
    tout->Branch("delta_R_zz",   &delta_R_zz);
    tout->Branch("delta_R_jj",   &delta_R_jj);
    tout->Branch("delta_R_zz_jj",   &delta_R_zz_jj);
    tout->Branch("mcweight", &weight);
    tout->Branch("delta_eta_jj", &delta_eta_jj);
    tout->Branch("centrality", &centrality);

    for(int n=0; n<(int)tin->GetEntries(); n++)
    {
	TLorentzVector ele[2], muon[2];
	TLorentzVector jet[2];
	TLorentzVector z[2];
	for (int i=0; i<2; i++)
	{
	    ele[i].SetPtEtaPhiE(0, 0, 0, 0);
	    muon[i].SetPtEtaPhiE(0, 0, 0, 0);
	    jet[i].SetPtEtaPhiE(0, 0, 0, 0);
	    z[i].SetPtEtaPhiE(0, 0, 0, 0);
	}

	TLorentzVector zz; zz.SetPtEtaPhiE(0, 0, 0, 0);
	TLorentzVector jj; jj.SetPtEtaPhiE(0, 0, 0, 0);

	double  delta_R_ll[2] = {-9999., -9999.};

	delta_eta_jj=-9999E10;
	centrality=-9999E10;
	delta_R_ll_1 = -9999.0;
	delta_R_ll_2 = -9999.0;
	delta_R_jj   = -9999.0;
	delta_R_zz   = -9999.0;
	delta_R_zz_jj= -9999.0;

	tin->GetEntry(n);

	int index_e=0, index_m=0, index_j=0;
	for (int i=0; i<10; i++)
	{
	    if(status[i] != 1) continue;
	    else if (abs(pdgID[i]) == 0 || abs(pdgID[i]) == 1 || abs(pdgID[i]) == 2 || abs(pdgID[i])==3 || abs(pdgID[i]) == 4 || abs(pdgID[i]) == 5 )
	    {
                jet[index_j].SetPxPyPzE(px[i], py[i], pz[i], E[i]);
		index_j++;
	    }
	    else if ( abs(pdgID[i]) == 11) // electron
	    {
		ele[index_e].SetPxPyPzE(px[i], py[i], pz[i], E[i]);
		index_e++;
	    }
	    else if ( abs(pdgID[i]) == 13) // muon
	    {
		muon[index_m].SetPxPyPzE(px[i], py[i], pz[i], E[i]);
		index_m++;
	    }
	}

	if ( index_e == 2 && index_m == 2 )
	{
	    z[0] = ele[0] + ele[1];
	    z[1] = muon[0] + muon[1];
	    delta_R_ll[0] = sqrt( pow( (ele[0].Eta() - ele[1].Eta()), 2) + pow( (ele[0].Phi() - ele[1].Phi()), 2) );
	    delta_R_ll[1] = sqrt( pow( (muon[0].Eta() - muon[1].Eta()), 2) + pow( (muon[0].Phi() - muon[1].Phi()), 2) );

	    zz = z[0] + z[1];
	    delta_R_zz = sqrt( pow( (z[0].Eta() - z[1].Eta()), 2) + pow( (z[0].Phi() - z[1].Phi()) , 2) );
	}
	else 
	{
	    cout << " event " << n << " doesn't have 4 leptons" << endl;
	    continue;
	}


	if( index_j == 2)
	{
	    jj = jet[0] + jet[1];
	    delta_R_jj = sqrt( pow( ( jet[0].Eta() - jet[1].Eta()), 2) + pow( (jet[0].Phi() - jet[1].Phi()), 2) );

	    delta_eta_jj = abs( jet[0].Eta() - jet[1].Eta() );

	    if( zz.E() > 1 )
	    {
		centrality = (zz.Rapidity() -( jet[0].Rapidity() + jet[1].Rapidity() )/2 ) / abs( jet[0].Rapidity() - jet[1].Rapidity() );
		delta_R_zz_jj = sqrt( pow((zz.Eta() - jj.Eta()), 2) + pow((zz.Phi() - jj.Phi()), 2) );
	    }
	}
	else 
	{
	    cout << "jet number less than 2" << endl;
	    continue;
	}


	l1_pt  = ele[0].Pt();
	l1_eta = ele[0].Eta();
	l1_phi = ele[0].Phi();
	l1_e   = ele[0].E();
	l2_pt  = ele[1].Pt();
	l2_eta = ele[1].Eta();
	l2_phi = ele[1].Phi();
	l2_e   = ele[1].E();
	l3_pt  = muon[0].Pt();
	l3_eta = muon[0].Eta();
	l3_phi = muon[0].Phi();
	l3_e   = muon[0].E();
	l4_pt  = muon[1].Pt();
	l4_eta = muon[1].Eta();
	l4_phi = muon[1].Phi();
	l4_e   = muon[1].E();
	leadingj_pt  = jet[0].Pt();
	leadingj_eta = jet[0].Eta();
	leadingj_phi = jet[0].Phi();
	leadingj_e   = jet[0].E();
	subleadingj_pt  = jet[1].Pt();
	subleadingj_eta = jet[1].Eta();
	subleadingj_phi = jet[1].Phi();
	subleadingj_e   = jet[1].E();
	z1_pt  = z[0].Pt();
	z1_eta = z[0].Eta();
	z1_phi = z[0].Phi();
	z1_e   = z[0].E();
	z1_m   = z[0].M();
	z2_pt  = z[1].Pt();
	z2_eta = z[1].Eta();
	z2_phi = z[1].Phi();
	z2_e   = z[1].E();
	z2_m   = z[1].M();
	zz_pt  = zz.Pt();
	zz_eta = zz.Eta();
	zz_phi = zz.Phi();
	zz_e   = zz.E();
	zz_m   = zz.M();
	jj_pt  = jj.Pt();
	jj_eta = jj.Eta();
	jj_phi = jj.Phi();
	jj_e   = jj.E();
	jj_m   = jj.M();
	delta_R_ll_1 = delta_R_ll[0];
	delta_R_ll_2 = delta_R_ll[1];

	tout->Fill();
    }
    fout->cd();
    tout->Write();

    time(&finish);
    double timedif = difftime(finish, start);

    cout << "time cost: " << timedif << endl;
}
